docs/tutorials/python/sagas/index.md

```diff
@@ -0,0 +1,551 @@
+        )
+        await asyncio.sleep(3)
+        raise RuntimeError("Flight service is down")
+
+    if "invalid" in input.book_flight_id:
+        raise Exception("Invalid flight booking, rolling back!")
+
+    print(f"Booking flight: {input.book_flight_id}")
+    return f"Booked flight: {input.book_flight_id}"
+```
+
+With the main booking Activities in place, it's time to define the compensation Activities.
+These undo actions are crucial for maintaining data consistency by rolling back successful steps if a subsequent step fails.
+
+### Define Compensation Activities
+
+For every action (`book_car`, `book_hotel`, and `book_flight`), you will create a corresponding undo action.
+These Activities will log the undo action and return a success message.
+
+```python
+@activity.defn
+async def undo_book_car(input: BookVacationInput) -> str:
+    print(f"Undoing booking of car: {input.book_car_id}")
+    return f"Undoing booking of car: {input.book_car_id}"
+
+
+@activity.defn
+async def undo_book_hotel(input: BookVacationInput) -> str:
+    print(f"Undoing booking of hotel: {input.book_hotel_id}")
+    return f"Undoing booking of hotel: {input.book_hotel_id}"
+
+
+@activity.defn
+async def undo_book_flight(input: BookVacationInput) -> str:
+    print(f"Undoing booking of flight: {input.book_flight_id}")
+    return f"Undoing booking of flight: {input.book_flight_id}"
+```
+
+For this example, if the number of attempts is less than the allowed number of attempts or if the booking ID is invalid, the Activity will raise exceptions to simulate failures, then run the corresponding undo action.
+
+By setting up these compensations, you'll ensure that your system can handle failures gracefully.
+Next, you'll focus on defining shared data classes and constants to support your Activities and Workflows.
+
+## Define Shared Data Classes and Constants
+
+Shared data classes and constants are used to pass data between Activities and Workflows.
+Common mistakes include using mutable data types such as lists or dictionaries, which can cause unexpected behavior.
+
+Also, Task Queues are shared resources that can be used by multiple Workflows and Workers.
+
+Create a new file named `shared.py`:
+
+```python
+from dataclasses import dataclass
+
+
+@dataclass
+class BookVacationInput:
+    book_user_id: str
+    book_car_id: str
+    book_hotel_id: str
+    book_flight_id: str
+    attempts: int
+
+
+TASK_QUEUE_NAME = "saga-task-queue"
+```
+
+These classes and constants will be used in both Activities, Workflows, and Workers.
+
+Workflows order and run the execution of Activities.
+With your Activities and shared data classes defined, the next step is to create the Workflow.
+This Workflow coordinates the execution of Activities and handle compensations to maintain consistency in case of failures.
+
+## Create the Workflow Definition
+
+In the context of Temporal Workflows, compensation refers to the actions taken to roll back a transaction if an error occurs.
```

Comment:
I would move this sentence up to the part where you first mention compensations.

Reply:
I partially agree although I think it's good to have it here too. I wouldn't move it so much as add a section to the introduction explaining how Temporal applications implement Sagas.

Reply:
Yeah but maybe reword/reemphasize this and move this to the first section when compensations are mentioned, because I think people would benefit from an explanation of what it means :)